Fireworks/extensions/database.py

```diff
@@ -186,22 +191,22 @@ def __repr__(self):
 
     return SimpleTable
 
-class DBSource(Source):
+class DBPipe(Pipe):
```

Comment:
If DBPipe is supposed to represent the query result shouldn't the actual query be performed in TablePipe? DBPipe should be agnostic of context and should take in just the query result. Because you're creating two sessions to do a query when you don't have to. It's unclear which session is actually performing the query.

Reply:
As far as I can tell, sessions and queries are tightly connected in sqlalchemy (https://docs.sqlalchemy.org/en/latest/orm/query.html). What's happening here, is you can make any SQLalchemy query on a TablePipe using the query method, and the result of that query will be used to make a DBPipe. The query object has a session, and that session also has its own associated query method. In theory, I could desession the DBPipe by manually setting query.session = None, and then requiring the user to provide a session, however, I think the approach here works fine. I can change it slightly to enforce that the session associated with a query is the same as self.session to remove ambiquity.
Both DBPipe and TablePipe have a session attribute, and manipulation of that session is separate from the methods that manipulate data. Hence, you can still swap out sessions easily and make your own decisions about how to manage them without having to mess with those methods.

Reply:
You only need to provide a session to the query object if you're instantiating yet directly (`Query(entities, session)`). You can also do `session.query(*args)`. which will make a query object with that session and is probably the preferred way.  It seems like the best way to handle this is instead of DBPipe taking in an engine, it takes in instead a session (that will have an engine associated anyway). That way when you query in TablePipe it passes in the TablePipe session into DBPipe. So instead of 
```DBPipe(self.table, self.engine, query, columns_and_types = columns_and_types)```
it's 
```DBPipe(self.table, self.session, query, columns_and_types = columns_and_types)```

This makes it more clear and better lines up the way sqlalchemy works at the ORM level

Reply:
I think I'm going to keep self.engine for now, because I realized that DBPipe attempts to infer column names from a table using table reflection if those are not provided by the user. This function requires an engine. 
https://docs.sqlalchemy.org/en/latest/core/reflection.html 
This is a convenience feature which lets you provide the name of a table as a string and a query, and the object will reflect the table out from the database and resolve the query. We can think about how to simplify this module and make it more streamlined and generalized, but I don't think it's a high priority right now because it works.

Reply:
maybe then make sessionmaker an additional input, because you don't need two sessionmakers (if it's used in conjunction with TablePipe). It also gives the user more control what kind of sessions they want to use (ex. Thread local sessions).